File: educational-core/src/com/jetbrains/edu/coursecreator/projectView/CCNode.java
package com.jetbrains.edu.coursecreator.projectView;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.projectView.ViewSettings;
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode;
import com.intellij.ide.util.treeView.AbstractTreeNode;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.jetbrains.edu.coursecreator.CCUtils;
import com.jetbrains.edu.learning.*;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.StudyItem;
import com.jetbrains.edu.learning.courseFormat.ext.CourseExt;
import com.jetbrains.edu.learning.projectView.DirectoryNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CCNode extends DirectoryNode {
  public CCNode(@NotNull Project project,
                PsiDirectory value,
                ViewSettings viewSettings) {
    super(project, value, viewSettings);
  }

  @Override
  public boolean canNavigate() {
    return true;
  }

  @Nullable
  @Override
  public AbstractTreeNode modifyChildNode(AbstractTreeNode childNode) {
    final AbstractTreeNode node = super.modifyChildNode(childNode);
    if (node != null) return node;
    Object value = childNode.getValue();
    if (value instanceof PsiElement) {
      PsiFile psiFile = ((PsiElement) value).getContainingFile();
      VirtualFile virtualFile = psiFile.getVirtualFile();

      Course course = StudyTaskManager.getInstance(myProject).getCourse();
      if (course == null) {
        return null;
      }
      EduConfigurator configurator = EduConfiguratorManager.forLanguage(course.getLanguageById());
      if (configurator == null) {
        return new CCStudentInvisibleFileNode(myProject, psiFile, myViewSettings);
      }
      if (EduUtils.isTaskDescriptionFile(virtualFile.getName())) {
        return null;
      }
      if (!EduUtils.isTestsFile(myProject, virtualFile)) {
        return new CCStudentInvisibleFileNode(myProject, psiFile, myViewSettings);
      } else {
        if (EduUtils.isConfiguredWithGradle(myProject) && CCUtils.isCourseCreator(myProject)) {
          return new CCStudentInvisibleFileNode(myProject, psiFile, myViewSettings);
        }
      }
    }
    return null;
  }

  @Override
  public PsiDirectoryNode createChildDirectoryNode(StudyItem item, PsiDirectory value) {
    return new CCNode(myProject, value, myViewSettings);
  }

  @Override
  protected void updateImpl(PresentationData data) {
    Project project = getProject();
    if (project != null && CCUtils.isCourseCreator(project) && EduUtils.isConfiguredWithGradle(project)) {
      PsiDirectory dir = getValue();
      VirtualFile directoryFile = dir.getVirtualFile();
      String name = directoryFile.getName();

      Course course = StudyTaskManager.getInstance(myProject).getCourse();
      // course is not null because of `CCUtils.isCourseCreator(project)` check above
      assert course != null;

      if (name.equals(CourseExt.getSourceDir(course)) || name.equals(CourseExt.getTestDir(course))) {
        data.setPresentableText(name);
        return;
      }
    }

    super.updateImpl(data);
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/projectView/CCNode.kt
package com.jetbrains.edu.coursecreator.projectView

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.projectView.ViewSettings
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode
import com.intellij.ide.util.treeView.AbstractTreeNode
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiElement
import com.intellij.ui.SimpleTextAttributes
import com.jetbrains.edu.coursecreator.CCUtils.isCourseCreator
import com.jetbrains.edu.coursecreator.courseignore.CourseIgnoreRules
import com.jetbrains.edu.learning.EduUtilsKt
import com.jetbrains.edu.learning.StudyTaskManager
import com.jetbrains.edu.learning.courseFormat.ext.configurator
import com.jetbrains.edu.learning.courseFormat.tasks.Task
import com.jetbrains.edu.learning.isTestsFile
import com.jetbrains.edu.learning.messages.EduCoreBundle
import com.jetbrains.edu.learning.projectView.DirectoryNode

class CCNode(
  project: Project,
  value: PsiDirectory,
  viewSettings: ViewSettings,
  private val courseIgnoreRules: CourseIgnoreRules,
  task: Task?
) : DirectoryNode(project, value, viewSettings, task) {

  private val excludedName: String?
  init {
    // show node as excluded, if it is not inside a task folder, but is in the .courseignore
    excludedName = if (task == null && courseIgnoreRules.isIgnored(value.virtualFile, project)) {
      EduCoreBundle.message("course.creator.course.view.excluded", value.name)
    }
    else {
      null
    }
  }

  override fun canNavigate(): Boolean = true

  override fun modifyChildNode(childNode: AbstractTreeNode<*>): AbstractTreeNode<*>? {
    val node = super.modifyChildNode(childNode)
    if (node != null) return node
    val value = childNode.value

    if (value is PsiDirectory) {
      return CCNode(myProject, value, settings, courseIgnoreRules, item)
    }
    if (value is PsiElement) {
      val psiFile = value.containingFile
      val virtualFile = psiFile.virtualFile
      val course = StudyTaskManager.getInstance(myProject).course ?: return null
      if (course.configurator == null) return CCStudentInvisibleFileNode(myProject, psiFile, settings, courseIgnoreRules)
      if (EduUtilsKt.isTaskDescriptionFile(virtualFile.name)) {
        return null
      }
      if (!virtualFile.isTestsFile(myProject)) {
        return CCStudentInvisibleFileNode(myProject, psiFile, settings, courseIgnoreRules)
      }
      else {
        if (isCourseCreator(myProject)) {
          return CCStudentInvisibleFileNode(myProject, psiFile, settings, courseIgnoreRules)
        }
      }
    }
    return null
  }

  override fun updateImpl(data: PresentationData) {
    super.updateImpl(data)
    if (excludedName != null) {
      data.clearText()
      data.addText(excludedName, SimpleTextAttributes.GRAY_ATTRIBUTES)
    }
  }

  override fun createChildDirectoryNode(value: PsiDirectory): PsiDirectoryNode {
    return CCNode(myProject, value, settings, courseIgnoreRules, item)
  }
}
